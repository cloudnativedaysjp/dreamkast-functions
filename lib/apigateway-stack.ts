import { Construct } from 'constructs';
import { Stack, StackProps, Fn,RemovalPolicy  } from 'aws-cdk-lib';
import { Function, IFunction } from 'aws-cdk-lib/aws-lambda';
import { Effect, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { ARecord, IHostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { ApiGatewayDomain } from 'aws-cdk-lib/aws-route53-targets';
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";

import * as apigateway from 'aws-cdk-lib/aws-apigateway';

import { ViewerCountSchema, ProfilePointSchema, ProfilePointsSchema, VoteSchema } from './schemas';
import { BuildConfig } from './build-config'

export interface APIGatewayProps extends StackProps {
    readonly certificate: Certificate,
    readonly hostedZone: IHostedZone,
    readonly lambda: {
        readonly voteCFP: IFunction,
        readonly postProfilePoint: IFunction,
        readonly getProfilePoint: IFunction,
    }
}

export class APIGatewayStack extends Stack {
    constructor(scope: Construct, id: string, props: APIGatewayProps, buildConfig: BuildConfig) {
        super(scope, id, props)

        // === [ API Gateway ] === 

        const api = new apigateway.RestApi(this, 'dkFunctionsApi',{
            restApiName: `dk-functions-${buildConfig.Environment}`,
            deployOptions: {
                stageName: 'v1',
                throttlingRateLimit: 60,
                throttlingBurstLimit: 3000,
                dataTraceEnabled: true,
                loggingLevel: apigateway.MethodLoggingLevel.INFO,
                accessLogDestination: new apigateway.LogGroupLogDestination(new LogGroup(this, 'ApiLogGroup', {
                    logGroupName: `${id}-apiGateway`,
                    retention: RetentionDays.ONE_MONTH,
                    removalPolicy: RemovalPolicy.DESTROY,
                })),
                accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields(),
            },
        });

        const requestValidator = new apigateway.RequestValidator(this, 'ApiRequestValidator', {
            restApi: api,
            validateRequestParameters: true,
            validateRequestBody: true,
        });

         // Custom Domain
        const domainName = new apigateway.DomainName(this, 'CustomDomain',{
            certificate: props.certificate,
            domainName: buildConfig.DomainName,
            endpointType: apigateway.EndpointType.REGIONAL,
        });
        domainName.addBasePathMapping(api,{
            basePath: '',
        });
        // Not used to follow the dk swagger specifications
        //domainName.addApiMapping(api.deploymentStage,{
        //    basePath: 'api/v1',
        //});

         // A Record
        new ARecord(this, 'APIARecod', {
            zone: props.hostedZone,
            recordName: buildConfig.DomainName,
            target: RecordTarget.fromAlias(new ApiGatewayDomain(domainName))
        });

         // EXECUTION ROLE
        const projectApiExecutionRole = new Role(this, 'ProjectApiExecutionRole', {
            assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
            inlinePolicies: {
            APIGWLambdaPolicy: new PolicyDocument({
                statements: [
                new PolicyStatement({
                    actions: ['lambda:Invoke*'],
                    effect: Effect.ALLOW,
                    resources: ['arn:aws:lambda:*:607167088920:function:*'],
                }),
                ]
            }),
            }
        });

        /* === [ RESOURCES ] === */
        
        const root = api.root;
        const apiv1 = root.addResource('api').addResource('v1')

        // Dreamkast
        new apigateway.ProxyResource(this,'ProxyResource',{
            parent: apiv1,
            anyMethod: true,
            defaultIntegration: new apigateway.HttpIntegration(`${buildConfig.DreamkastApiBaseUrl}/api/v1/{proxy}`,{
                    proxy: true,
                    httpMethod: 'ANY',
                    options:{
                        credentialsPassthrough: true,
                        requestParameters: {
                            "integration.request.path.proxy": "method.request.path.proxy"
                        },
                    },
                }),
            defaultMethodOptions: {
                requestParameters:{
                    "method.request.path.proxy": true,
                }
            },
        });


        // TRACKS
        const tracks = apiv1.addResource('tracks');
        const trackid = tracks.addResource('{trackId}');
        const viewerCount = trackid.addResource('viewer_count', {
            defaultCorsPreflightOptions: {
                statusCode: 200,
                allowOrigins: apigateway.Cors.ALL_ORIGINS,
            }
        });

        // TALKS
        const talks = apiv1.addResource('talks');
        const talkId = talks.addResource('{talkId}')
        const vote = talkId.addResource('vote');

        // Profile
        const profiles = apiv1.addResource('profile');
        const profileId = profiles.addResource('{profileId}');
        const point = profileId.addResource('point');
        const points = profileId.addResource('points');

        /* === [   MODEL   ] === */

        const viewerCountModel = api.addModel('viewerCountModel',{
            contentType: 'application/json',
            modelName: 'ViewerCount',
            schema: ViewerCountSchema,
        })

        const voteModel = api.addModel('voteModel',{
            contentType: 'application/json',
            modelName: 'Vote',
            schema: VoteSchema,
        })

        const profilePointModel = api.addModel('profilePointModel',{
            contentType: 'application/json',
            modelName: 'ProfilePoint',
            schema: ProfilePointSchema,
        })
        
        const profilePointsModel = api.addModel('profilePointsModel',{
            contentType: 'application/json',
            modelName: 'ProfilePoints',
            schema: ProfilePointsSchema,
        })

        /* === [   ResponseParameters   ] === */

        const CorsResponseParameters = {
            'method.response.header.Access-Control-Allow-Methods': "'GET,POST'",
            'method.response.header.Access-Control-Allow-Origin': buildConfig.AccessControlAllowOrigin,
            'method.response.header.Access-Control-Allow-Headers': "'Content-Type,Authorization,X-Amz-Date,X-Api-Key,X-Amz-Security-Token'",
        }
        const CorsMethodResponseParameters = {
            'method.response.header.Access-Control-Allow-Methods': true,
            'method.response.header.Access-Control-Allow-Origin': true,
            'method.response.header.Access-Control-Allow-Headers': true,
        }

        /* === [   IntegrationResponse   ] === */

        const integrationResponse200 = {
            statusCode: '200',
            responseParameters: CorsResponseParameters,
        }
        const integrationResponse400 = {
            statusCode: '400',
            selectionPattern: 'Error400:.*',
            responseParameters: CorsResponseParameters,
            responseTemplates: {
                'application/json': JSON.stringify({"message":"bad request"}),
            },
        }
        const integrationResponse404 = {
            statusCode: '404',
            selectionPattern: 'Error404:.*',
            responseParameters: CorsResponseParameters,
            responseTemplates: {
                'application/json': JSON.stringify({"message":"not found"}),
            },
        }
        const integrationResponse500 = {
            statusCode: '500',
            selectionPattern: 'Error500:.*',
            responseParameters: CorsResponseParameters,
            responseTemplates: {
                'application/json': JSON.stringify({"message":"internal server error"}),
            },
        }

        /* === [   MethodResponse   ] === */

        const methodResponses200 = {
            statusCode: '200',
            responseParameters: CorsMethodResponseParameters,
            responseModels: {
                'application/json': apigateway.Model.EMPTY_MODEL,
            },
        }
        const methodResponses400 = {
            statusCode: '400',
            responseParameters: CorsMethodResponseParameters,
            responseModels: {
                'application/json': apigateway.Model.ERROR_MODEL,
            },
        }
        const methodResponses404 = {
            statusCode: '404',
            responseParameters: CorsMethodResponseParameters,
            responseModels: {
                'application/json': apigateway.Model.ERROR_MODEL,
            },
        }
        const methodResponses500 = {
            statusCode: '500',
            responseParameters: CorsMethodResponseParameters,
            responseModels: {
                'application/json': apigateway.Model.ERROR_MODEL,
            },
        }

        /* === [   METHODS   ] === */

        // GET /tracks/{trackID}/viewer_count -> GetViewerCountFunction

        const getViewerCountFunctionArn = Fn.importValue(`viewerCountStack-GetViewerCountFunction-Arn-${buildConfig.Environment}`);
        const getViewerCountFunction = Function.fromFunctionArn(this, 'GetViewerCountFunction', getViewerCountFunctionArn);
        viewerCount.addMethod('GET',
            // Integration
            new apigateway.LambdaIntegration( getViewerCountFunction,
                {
                    proxy: false,
                    credentialsRole: projectApiExecutionRole,
                    passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
                    requestTemplates: {
                        'application/json': `{
                            #set($trackId = $util.escapeJavaScript($input.params().get("path").get("trackId")))
                            "trackId": #if($trackId == "" ) 0 #{else} "$trackId" #end
                        }`,
                    },
                    integrationResponses: [
                        integrationResponse200,
                        integrationResponse400,
                        integrationResponse404,
                    ],
                },
            ),
            // MethodOptions
            {
                requestParameters: {
                    "method.request.path.trackId": true,
                },
                requestValidator: requestValidator,
                methodResponses: [
                    {
                        statusCode: '200',
                        responseParameters: CorsMethodResponseParameters,
                        responseModels: {
                            // Need ?
                            'application/json': viewerCountModel,
                        },
                    },
                    methodResponses400,
                    methodResponses404,
                ]
            },
        );

        // POST /talk/{trackID}/vote -> VoteCFP
        vote.addMethod('POST',
            // Integration
            new apigateway.LambdaIntegration( props.lambda.voteCFP,
                {
                    proxy: false,
                    credentialsRole: projectApiExecutionRole,
                    passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
                    requestTemplates: {
                        'application/json': `{
                            "eventAbbr":"$util.escapeJavaScript($input.path('$').eventAbbr)",
                            "talkId": "$util.escapeJavaScript($input.params().get("path").get("talkId"))",
                            "globalIp": "$util.escapeJavaScript($context.identity.sourceIp)"
                        }`,
                    },
                    integrationResponses: [
                        integrationResponse200,
                        integrationResponse400,
                        integrationResponse404,
                    ],
                },
            ),
            // MethodOptions
            {
                requestValidator: requestValidator,
                requestParameters: {
                    "method.request.path.talkId": true,
                },
                requestModels: {
                    'application/json': voteModel,
                },
                methodResponses: [
                    methodResponses200,
                    methodResponses400,
                    methodResponses404,
                ]
            },
        );

        // POST /profiles/{profileId}/point -> PostProfilePointFunction
        point.addMethod('POST',
            // Integration
            new apigateway.LambdaIntegration( props.lambda.postProfilePoint,
                {
                    proxy: false,
                    credentialsRole: projectApiExecutionRole,
                    passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
                    requestTemplates: {
                        'application/json': `{
                            "conference":"$util.escapeJavaScript($input.path('$').eventAbbr)",
                            "profileId":"$util.escapeJavaScript($input.params().get("path").get("profileId"))",
                            "point": "$util.escapeJavaScript($input.path('$').point)",
                            "reasonId":"$util.escapeJavaScript($input.path('$').reasonId)"
                        }`,
                    },
                    integrationResponses: [
                        integrationResponse200,
                        integrationResponse400,
                        integrationResponse404,
                    ],
                },
            ),
            // MethodOptions
            {
                requestValidator: requestValidator,
                requestParameters: {
                    "method.request.path.profileId": true,
                },
                requestModels: {
                    'application/json': profilePointModel,
                },
                methodResponses: [
                    methodResponses200,
                    methodResponses400,
                    methodResponses404,
                    methodResponses500,
                ]
            },
        );

        // GET /profiles/{profileId}/points -> GetProfilePointFunction
        points.addMethod('GET',
        // Integration
        new apigateway.LambdaIntegration( props.lambda.getProfilePoint,
            {
                proxy: false,
                credentialsRole: projectApiExecutionRole,
                passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
                requestTemplates: {
                    'application/json': `{
                        "profileId":"$util.escapeJavaScript($input.params().get("path").get("profileId"))",
                        "conference":"$util.escapeJavaScript($input.params("eventAbbr"))"
                    }`,
                },
                integrationResponses: [
                    integrationResponse200,
                    integrationResponse400,
                    integrationResponse500,
                ],
            },
        ),
        // MethodOptions
        {
            requestValidator: requestValidator,
            requestParameters: {
                "method.request.querystring.eventAbbr": false,
                "method.request.path.profileId": true,
            },
            methodResponses: [
                {
                    statusCode: '200',
                    responseParameters: CorsMethodResponseParameters,
                    responseModels: {
                        'application/json': profilePointsModel,
                    },
                },
                methodResponses400,
                methodResponses500,
            ]
        },
    );
    }
}