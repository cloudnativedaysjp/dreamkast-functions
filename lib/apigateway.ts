import { Construct } from 'constructs'
import { Stack, StackProps, Fn, RemovalPolicy } from 'aws-cdk-lib'
import { Function, IFunction } from 'aws-cdk-lib/aws-lambda'
import {
  Effect,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from 'aws-cdk-lib/aws-iam'
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager'
import { ARecord, IHostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53'
import { ApiGatewayDomain } from 'aws-cdk-lib/aws-route53-targets'
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs'

import * as apigateway from 'aws-cdk-lib/aws-apigateway'

import {
  ViewerCountSchema,
  ProfilePointRequestSchema,
  ProfilePointsResponseSchema,
  VoteSchema,
  DkUiDataSchema,
  DkUiDataMutationSchema,
} from './schemas'
import { BuildConfig } from './buildConfig'

const requestPassThroughTemplate = `
#set($allParams = $input.params())
{
  "body" : $input.json('$'),
  #foreach($type in $allParams.keySet())
  #set($params = $allParams.get($type))
  "$type" : {
    #foreach($paramName in $params.keySet())
    "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
    #if($foreach.hasNext),#end
    #end
  },
  #end
  "context" : {
    "sourceIp" : "$context.identity.sourceIp"
  }
}
`

export interface APIGatewayProps extends StackProps {
  readonly certificate: Certificate
  readonly hostedZone: IHostedZone
  readonly lambda: {
    readonly getViewerCount: IFunction
    readonly voteCFP: IFunction
    readonly postProfilePoint: IFunction
    readonly getProfilePoint: IFunction
    readonly getDkUiData: IFunction
    readonly postDkUiData: IFunction
  }
}

export function newAPIGatewayResources(
  scope: Construct,
  id: string,
  props: APIGatewayProps,
  buildConfig: BuildConfig,
) {
  // === [ API Gateway ] ===

  const api = new apigateway.RestApi(scope, 'dkFunctionsApi', {
    restApiName: `dk-functions-${buildConfig.Environment}`,
    deployOptions: {
      stageName: 'v1',
      throttlingRateLimit: 60,
      throttlingBurstLimit: 3000,
      dataTraceEnabled: true,
      loggingLevel: apigateway.MethodLoggingLevel.INFO,
      accessLogDestination: new apigateway.LogGroupLogDestination(
        new LogGroup(scope, 'ApiLogGroup', {
          logGroupName: `${id}-apiGateway-${buildConfig.Environment}`,
          retention: RetentionDays.ONE_MONTH,
          removalPolicy: RemovalPolicy.DESTROY,
        }),
      ),
      accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields(),
    },
  })

  const requestValidator = new apigateway.RequestValidator(
    scope,
    'ApiRequestValidator',
    {
      restApi: api,
      validateRequestParameters: true,
      validateRequestBody: true,
    },
  )

  // Custom Domain
  const domainName = new apigateway.DomainName(scope, 'CustomDomain', {
    certificate: props.certificate,
    domainName: buildConfig.DomainName,
    endpointType: apigateway.EndpointType.REGIONAL,
  })
  domainName.addBasePathMapping(api, {
    basePath: '',
  })
  // Not used to follow the dk swagger specifications
  //domainName.addApiMapping(api.deploymentStage,{
  //    basePath: 'api/v1',
  //});

  // A Record
  new ARecord(scope, 'APIARecod', {
    zone: props.hostedZone,
    recordName: buildConfig.DomainName,
    target: RecordTarget.fromAlias(new ApiGatewayDomain(domainName)),
  })

  // EXECUTION ROLE
  const projectApiExecutionRole = new Role(scope, 'ProjectApiExecutionRole', {
    assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
    inlinePolicies: {
      APIGWLambdaPolicy: new PolicyDocument({
        statements: [
          new PolicyStatement({
            actions: ['lambda:Invoke*'],
            effect: Effect.ALLOW,
            resources: ['arn:aws:lambda:*:607167088920:function:*'],
          }),
        ],
      }),
    },
  })

  /* === [ RESOURCES ] === */

  const root = api.root
  const apiv1 = root.addResource('api').addResource('v1', {
    defaultCorsPreflightOptions: {
      statusCode: 200,
      allowOrigins: [buildConfig.AccessControlAllowOrigin],
      allowMethods: apigateway.Cors.ALL_METHODS,
      allowHeaders: apigateway.Cors.DEFAULT_HEADERS,
    },
  })

  // Dreamkast
  new apigateway.ProxyResource(scope, 'ProxyResource', {
    parent: apiv1,
    anyMethod: true,
    defaultIntegration: new apigateway.HttpIntegration(
      `${buildConfig.DreamkastApiBaseUrl}/api/v1/{proxy}`,
      {
        proxy: true,
        httpMethod: 'ANY',
        options: {
          credentialsPassthrough: true,
          requestParameters: {
            'integration.request.path.proxy': 'method.request.path.proxy',
          },
        },
      },
    ),
    defaultMethodOptions: {
      requestParameters: {
        'method.request.path.proxy': true,
      },
    },
  })

  // TRACKS
  const tracks = apiv1.addResource('tracks')
  const trackid = tracks.addResource('{trackId}')
  const viewerCount = trackid.addResource('viewer_count')

  // TALKS
  const talks = apiv1.addResource('talks')
  const talkId = talks.addResource('{talkId}')
  const vote = talkId.addResource('vote')

  // Profile
  const profiles = apiv1.addResource('profile')
  const profileId = profiles.addResource('{profileId}')
  const point = profileId.addResource('point')
  const points = profileId.addResource('points')

  // dk-ui AppData
  const appData = apiv1
    .addResource('app-data')
    .addResource('{profileId}')
    .addResource('conference')
    .addResource('{conference}')

  /* === [   MODEL   ] === */

  const viewerCountModel = api.addModel('viewerCountModel', {
    contentType: 'application/json',
    modelName: 'ViewerCount',
    schema: ViewerCountSchema,
  })

  const voteModel = api.addModel('voteModel', {
    contentType: 'application/json',
    modelName: 'Vote',
    schema: VoteSchema,
  })

  const profilePointModel = api.addModel('profilePointModel', {
    contentType: 'application/json',
    modelName: 'ProfilePoint',
    schema: ProfilePointRequestSchema,
  })

  const profilePointsModel = api.addModel('profilePointsModel', {
    contentType: 'application/json',
    modelName: 'ProfilePoints',
    schema: ProfilePointsResponseSchema,
  })

  const dkUiDataModel = api.addModel('dkUiDataModel', {
    contentType: 'application/json',
    modelName: 'DkUiData',
    schema: DkUiDataSchema,
  })

  const dkUiDataMutationModel = api.addModel('dkUiDataMutationModel', {
    contentType: 'application/json',
    modelName: 'DkUiDataMutation',
    schema: DkUiDataMutationSchema,
  })

  /* === [   ResponseParameters   ] === */

  const CorsResponseParameters = {
    'method.response.header.Access-Control-Allow-Methods': "'GET,POST'",
    'method.response.header.Access-Control-Allow-Origin': `'${buildConfig.AccessControlAllowOrigin}'`,
    'method.response.header.Access-Control-Allow-Headers':
      "'Content-Type,Authorization,X-Amz-Date,X-Api-Key,X-Amz-Security-Token'",
  }
  const CorsMethodResponseParameters = {
    'method.response.header.Access-Control-Allow-Methods': true,
    'method.response.header.Access-Control-Allow-Origin': true,
    'method.response.header.Access-Control-Allow-Headers': true,
  }

  /* === [   IntegrationResponse   ] === */

  const integrationResponse200 = {
    statusCode: '200',
    responseParameters: CorsResponseParameters,
  }
  const integrationResponse400 = {
    statusCode: '400',
    selectionPattern: 'Error400:.*',
    responseParameters: CorsResponseParameters,
    responseTemplates: {
      'application/json': JSON.stringify({ message: 'bad request' }),
    },
  }
  const integrationResponse404 = {
    statusCode: '404',
    selectionPattern: 'Error404:.*',
    responseParameters: CorsResponseParameters,
    responseTemplates: {
      'application/json': JSON.stringify({ message: 'not found' }),
    },
  }
  const integrationResponse500 = {
    statusCode: '500',
    selectionPattern: 'Error500:.*',
    responseParameters: CorsResponseParameters,
    responseTemplates: {
      'application/json': JSON.stringify({
        message: 'internal server error',
      }),
    },
  }

  /* === [   MethodResponse   ] === */

  const methodResponses200 = {
    statusCode: '200',
    responseParameters: CorsMethodResponseParameters,
    responseModels: {
      'application/json': apigateway.Model.EMPTY_MODEL,
    },
  }
  const methodResponses400 = {
    statusCode: '400',
    responseParameters: CorsMethodResponseParameters,
    responseModels: {
      'application/json': apigateway.Model.ERROR_MODEL,
    },
  }
  const methodResponses404 = {
    statusCode: '404',
    responseParameters: CorsMethodResponseParameters,
    responseModels: {
      'application/json': apigateway.Model.ERROR_MODEL,
    },
  }
  const methodResponses500 = {
    statusCode: '500',
    responseParameters: CorsMethodResponseParameters,
    responseModels: {
      'application/json': apigateway.Model.ERROR_MODEL,
    },
  }

  /* === [   METHODS   ] === */

  // GET /tracks/{trackID}/viewer_count -> GetViewerCountFunction
  viewerCount.addMethod(
    'GET',
    // Integration
    new apigateway.LambdaIntegration(props.lambda.getViewerCount, {
      proxy: false,
      credentialsRole: projectApiExecutionRole,
      passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': requestPassThroughTemplate,
      },
      integrationResponses: [
        integrationResponse200,
        integrationResponse400,
        integrationResponse404,
      ],
    }),
    // MethodOptions
    {
      requestParameters: {
        'method.request.path.trackId': true,
      },
      requestValidator: requestValidator,
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: CorsMethodResponseParameters,
          responseModels: {
            // Need ?
            'application/json': viewerCountModel,
          },
        },
        methodResponses400,
        methodResponses404,
      ],
    },
  )

  // POST /talk/{trackID}/vote -> VoteCFP
  vote.addMethod(
    'POST',
    // Integration
    new apigateway.LambdaIntegration(props.lambda.voteCFP, {
      proxy: false,
      credentialsRole: projectApiExecutionRole,
      passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': requestPassThroughTemplate,
      },
      integrationResponses: [
        integrationResponse200,
        integrationResponse400,
        integrationResponse404,
      ],
    }),
    // MethodOptions
    {
      requestValidator: requestValidator,
      requestParameters: {
        'method.request.path.talkId': true,
      },
      requestModels: {
        'application/json': voteModel,
      },
      methodResponses: [
        methodResponses200,
        methodResponses400,
        methodResponses404,
      ],
    },
  )

  // POST /profiles/{profileId}/point -> PostProfilePointFunction
  point.addMethod(
    'POST',
    // Integration
    new apigateway.LambdaIntegration(props.lambda.postProfilePoint, {
      proxy: false,
      credentialsRole: projectApiExecutionRole,
      passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': requestPassThroughTemplate,
      },
      integrationResponses: [
        integrationResponse200,
        integrationResponse400,
        integrationResponse404,
      ],
    }),
    // MethodOptions
    {
      requestValidator: requestValidator,
      requestParameters: {
        'method.request.path.profileId': true,
      },
      requestModels: {
        'application/json': profilePointModel,
      },
      methodResponses: [
        methodResponses200,
        methodResponses400,
        methodResponses404,
        methodResponses500,
      ],
    },
  )

  // GET /profiles/{profileId}/points -> GetProfilePointFunction
  points.addMethod(
    'GET',
    // Integration
    new apigateway.LambdaIntegration(props.lambda.getProfilePoint, {
      proxy: false,
      credentialsRole: projectApiExecutionRole,
      passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': requestPassThroughTemplate,
      },
      integrationResponses: [
        integrationResponse200,
        integrationResponse400,
        integrationResponse500,
      ],
    }),
    // MethodOptions
    {
      requestValidator: requestValidator,
      requestParameters: {
        'method.request.querystring.conference': true,
        'method.request.path.profileId': true,
      },
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: CorsMethodResponseParameters,
          responseModels: {
            'application/json': profilePointsModel,
          },
        },
        methodResponses400,
        methodResponses500,
      ],
    },
  )

  // GET /app-data/{profileId}/conference/{conference}
  appData.addMethod(
    'GET',
    // Integration
    new apigateway.LambdaIntegration(props.lambda.getDkUiData, {
      proxy: false,
      credentialsRole: projectApiExecutionRole,
      passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': requestPassThroughTemplate,
      },
      integrationResponses: [
        integrationResponse200,
        integrationResponse400,
        integrationResponse500,
      ],
    }),
    // MethodOptions
    {
      requestValidator: requestValidator,
      requestParameters: {
        'method.request.path.conference': true,
        'method.request.path.profileId': true,
      },
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: CorsMethodResponseParameters,
          responseModels: {
            'application/json': dkUiDataModel,
          },
        },
        methodResponses400,
        methodResponses500,
      ],
    },
  )

  // POST /app-data/{profileId}/conference/{conference}
  appData.addMethod(
    'POST',
    // Integration
    new apigateway.LambdaIntegration(props.lambda.postDkUiData, {
      proxy: false,
      credentialsRole: projectApiExecutionRole,
      passthroughBehavior: apigateway.PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': requestPassThroughTemplate,
      },
      integrationResponses: [
        integrationResponse200,
        integrationResponse400,
        integrationResponse500,
      ],
    }),
    // MethodOptions
    {
      requestValidator: requestValidator,
      requestParameters: {
        'method.request.path.conference': true,
        'method.request.path.profileId': true,
      },
      requestModels: {
        'application/json': dkUiDataMutationModel,
      },
      methodResponses: [
        methodResponses200,
        methodResponses400,
        methodResponses500,
      ],
    },
  )

  return {}
}
