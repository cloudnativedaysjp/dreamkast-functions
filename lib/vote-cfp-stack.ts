import { Construct } from 'constructs';
import { Stack, StackProps, RemovalPolicy } from 'aws-cdk-lib';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { AttributeType, BillingMode, Table } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { BuildConfig } from './build-config'
import { IFunction } from 'aws-cdk-lib/aws-lambda';


export class VoteCFPStack extends Stack {
    public readonly voteCFPFunction: IFunction 

    constructor(scope: Construct, id: string, props: StackProps, buildConfig: BuildConfig) {
        super(scope, id, props)

        // Dynamo DB

        const voteTable = new Table(this, 'VoteTable', {
            billingMode: BillingMode.PAY_PER_REQUEST,
            partitionKey: { name: 'eventName', type: AttributeType.STRING },
            sortKey: { name: 'timestamp', type: AttributeType.NUMBER },
            removalPolicy: RemovalPolicy.RETAIN,
        });

        // Lambda: Vote

        this.voteCFPFunction = new NodejsFunction(this, 'voteCFP',{
            entry: 'src/vote_cfp.ts',
            environment: {
                TABLENAME: voteTable.tableName,
            },
        });
        this.voteCFPFunction.addToRolePolicy(new PolicyStatement({
            resources: [
                voteTable.tableArn,
            ],
            actions: [
                'dynamodb:PutItem',
            ]
        }));
    }
}

