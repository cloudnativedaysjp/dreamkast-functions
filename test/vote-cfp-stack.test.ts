import * as cdk from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { VoteCFPStack } from '../lib/vote-cfp-stack'

test('snapshot test', () => {
  const app = new cdk.App()

  const stack = new VoteCFPStack(
    app,
    `stack`,
    {
      stackName: `test`,
      env: {
        region: 'ap-northeast-1',
      },
    },
    {
      DreamkastApiBaseUrl: 'https://staging.dev.cloudnativedays.jp',
      AWSProfileRegion: 'us-east-2',
      Environment: 'test',
      DomainName: 'api.test.cloudnativedays.jp',
      HostedZoneID: 'test',
      ZoneName: 'cloudnativedays.jp',
      AccessControlAllowOrigin: 'hoge.com',
      GetTracksURL: 'https://test.cloudnativedays.jp',
    },
  )

  // スタックからテンプレート(JSON)を生成
  const template = Template.fromStack(stack).toJSON()

  // 生成したテンプレートとスナップショットが同じか検証
  expect(template).toMatchSnapshot()
})
