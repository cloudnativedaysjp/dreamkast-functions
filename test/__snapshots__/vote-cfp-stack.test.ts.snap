// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "VoteTableC0BC27A7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "eventAbbr",
            "AttributeType": "S",
          },
          {
            "AttributeName": "timestamp",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "eventAbbr",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "timestamp",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "voteCFP7341FB29": {
      "DependsOn": [
        "voteCFPServiceRoleDefaultPolicy49FDC81F",
        "voteCFPServiceRole6841857B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "[HASH REMOVED]",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLENAME": {
              "Ref": "VoteTableC0BC27A7",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "voteCFPServiceRole6841857B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "voteCFPServiceRole6841857B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "voteCFPServiceRoleDefaultPolicy49FDC81F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:PutItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "VoteTableC0BC27A7",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "voteCFPServiceRoleDefaultPolicy49FDC81F",
        "Roles": [
          {
            "Ref": "voteCFPServiceRole6841857B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
